Zend/zend_operators.h

```diff
@@ -919,6 +919,58 @@ static zend_always_inline void zend_unwrap_reference(zval *op) /* {{{ */
 }
 /* }}} */
 
+static zend_always_inline bool zend_strnieq(const char *ptr1, const char *ptr2, size_t num)
+{
+	const char *end = ptr1 + num;
+	while (ptr1 < end) {
+		if (zend_tolower_ascii(*ptr1++) != zend_tolower_ascii(*ptr2++)) {
+			return 0;
+		}
+	}
+	return 1;
+}
+
+static zend_always_inline const char *
+zend_memnistr(const char *haystack, const char *needle, size_t needle_len, const char *end)
+{
+	ZEND_ASSERT(end >= haystack);
+
+	if (UNEXPECTED(needle_len == 0)) {
+		return haystack;
+	}
+
+	const char first_lower = zend_tolower_ascii(*needle);
+	const char first_upper = zend_toupper_ascii(*needle);
+	const char *p_lower = (const char *)memchr(haystack, first_lower, end - haystack);
```

Comment:
For future scope: I think there may be a way to optimize this with SSE for really long strings, but it may not be worth it, especially with stripos not being used that often.

(both checking if the first character exists case-insensitively, and checking if long needles match)

**Definitely not something to add to this PR.**

`(*needle_byte ^ *haystack_byte) == *needle_mask`, where needle_mask is 0xff or 0xcf (~0x20), since `'a' ^ 'A'` is 0x20)

For a block of 8/16/32 bytes, that could be checked by:

1. Computing the needle and mask and extending it to 8 bytes
2. Bitwise xoring the needle block with the haystack mask
3. Masking the xored block (0xff or 0xcf, depending on whether this was an alphabet character. `0xff ^ first_lower ^ first_upper`)
4. Checking for 0 bytes in the mask
5. Combining to a word
6. Checking if the combination was non-zero
7. Searching in that block normally for the first character


This would make the worst case of case-insensitive search faster (`stripos(str_repeat('A', 1000000), 'a')`) by stopping at the first lowercase or uppercase byte . The memchr could also start at whatever that result was if `needle_len > 1`.

Reply:
I know nothing about SSE, but as you mentioned, the new implementation should already be much better for large strings, we can always improve it at a later point.